Refactoring in prepation of docstring clean-up and extension

mseeger opened the pull request:
*Issue #, if available:*

*Description of changes:*


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

jgolebiowski commented on syne_tune/optimizer/schedulers/searchers/kde/kde_searcher.py:
```diff
@@ -95,13 +93,13 @@ def __init__(
         config_space: dict,
         metric: str,
         points_to_evaluate: Optional[List[dict]] = None,
-        mode: str = "min",
-        num_min_data_points: int = None,
-        top_n_percent: int = 15,
-        min_bandwidth: float = 1e-3,
-        num_candidates: int = 64,
-        bandwidth_factor: int = 3,
-        random_fraction: float = 0.33,
```
Same as syne_tune/optimizer/schedulers/searchers/bore/bore.py - I think this is more confusing now - users wont know the default values just by looking at the API but will need to dive into the implementation.

jgolebiowski replied:
NOTE: as long as the default values are in the docstring it makes sense to me.

jgolebiowski commented on syne_tune/optimizer/schedulers/searchers/kde/multi_fidelity_kde_searcher.py:
```diff
@@ -81,16 +79,18 @@ def __init__(
         config_space: Dict,
         metric: str,
         points_to_evaluate: Optional[List[Dict]] = None,
-        mode: str = "min",
-        num_min_data_points: int = None,
-        top_n_percent: int = 15,
-        min_bandwidth: float = 0.1,
-        num_candidates: int = 64,
-        bandwidth_factor: int = 3,
-        random_fraction: float = 0.33,
```
Same as syne_tune/optimizer/schedulers/searchers/bore/bore.py L40

jgolebiowski reviewed (commented):
The docstring chnages look good, I have added some comments to the code changes but they are mostly minor.

mseeger commented on syne_tune/backend/trial_backend.py:
```diff
@@ -126,15 +139,21 @@ def pause_trial(self, trial_id: int, result: Optional[dict] = None):
-        Backend specific operation that pauses the trial.
-        """
-        raise NotImplementedError()
+        raise NotImplementedError
```
we've been using abc in the past, but it has quite some issues, for example with serialization, so we removed it. It is one of these things that do not add value.

jgolebiowski replied:
I see, if you have seen problems with serialization of ABC's classes then lets leave it like this.

mseeger commented on syne_tune/callbacks/tensorboard_callback.py:
```diff
@@ -0,0 +1,180 @@
+        :param mode: Determined whether we maximize ("max") or minimize ("min")
+            the target metric.
+        """
+        if mode is None:
+            mode = "min"
```
I think it is best practice to define default values of base class methods in the code (so always use = None), otherwise they need to be copy and pasted to subclasses, and this introduces errors. Also, strings are mutable.

jgolebiowski replied:
I see your point about inheritance, as long as we keep the default in the docstrings then I am ok with that. 

I think the best practice are arguable, see for example an implementation of Adam in tensorflow https://github.com/keras-team/keras/blob/v2.10.0/keras/optimizers/optimizer_v2/adam.py#L26-L270 or sgd in Pytorch https://pytorch.org/docs/stable/_modules/torch/optim/sgd.html#SGD which are(IMHO) a relatively well written library 

As a note, strings in python are immutable, see second sentence of https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str

mseeger replied:
I just checked the sphinx docs, and they recommend to mention type and default in the docstring, something we are not right now doing. In the second PR, I will do that for all classes which are user-facing. This way, we get all of that info also in the API docs.

mseeger commented on syne_tune/callbacks/tensorboard_callback.py:
```diff
@@ -0,0 +1,180 @@
+        :param target_metric: Defines the metric we aim to optimize. If this
+            argument is set, we report the cumulative optimum of this metric as
+            well as the optimal hyperparameters we have found so far.
+        :param mode: Determined whether we maximize ("max") or minimize ("min")
```
OK.

mseeger commented on syne_tune/callbacks/tensorboard_callback.py:
```diff
@@ -0,0 +1,180 @@
+            "metric_sign": self.metric_sign,
+            "output_path": self.output_path,
+        }
+        return state
```
I don't think this changes anything, as `state` is just been created here. I did not write this code.

jgolebiowski replied:
Good point, I was thinking about the version with 
```
del state['writer']
return state
```
which is confusing. The current version is good.

mseeger commented on syne_tune/callbacks/tensorboard_callback.py:
```diff
@@ -0,0 +1,180 @@
+            f"tensorboard --logdir {self.output_path}"
+        )
+
+    def __getstate__(self):
```
OK

mseeger commented on syne_tune/callbacks/tensorboard_callback.py:
```diff
@@ -0,0 +1,180 @@
+        }
+        return state
+
+    def __setstate__(self, state):
```
That sounds a bit unsafe to me.

jgolebiowski replied:
I see the concern, in this case I am happy with the current state.

mseeger commented on syne_tune/experiments.py:
```diff
@@ -184,79 +194,139 @@ def load_experiment(
+    :return: Dictionary from tuner name to metadata dict
     """
-    res = {}
+    path_filter = _impute_path_filter(path_filter)
```
Use of these helper functions seems cleaner to me, as this is needed several times. Also, PyCharm tells me not to use lambda, but define the function directly.

mseeger replied:
I actually do not like lambda functions a lot, try not to use them.

mseeger commented on syne_tune/experiments.py:
```diff
@@ -184,79 +194,139 @@ def load_experiment(
+ExperimentFilter = Callable[[ExperimentResult], bool]
+
+
+def _impute_path_filter(pfilter: PathFilter) -> PathFilter:
+    if pfilter is None:
+
+        def pfilter(path) -> bool:
+            return True
+
+    return pfilter
+
+
+def _impute_experiment_filter(efilter: ExperimentFilter) -> ExperimentFilter:
+    if efilter is None:
+
+        def efilter(result) -> bool:
+            return True
+
+    return efilter
```
OK

mseeger commented on syne_tune/experiments.py:
```diff
@@ -184,79 +194,139 @@ def load_experiment(
+    :param load_tuner: Whether to load the tuner in addition to metadata and results
+    :return: List of `ExperimentResult` objects
+    """
+    path_filter = _impute_path_filter(path_filter)
```
OK

mseeger commented on syne_tune/experiments.py:
```diff
@@ -184,79 +194,139 @@ def load_experiment(
+    :return: List of `ExperimentResult` objects
+    """
+    path_filter = _impute_path_filter(path_filter)
+    if experiment_filter is None:
+
+        def experiment_filter(result) -> bool:
+            return True
```
OK

mseeger commented on syne_tune/optimizer/schedulers/fifo.py:
```diff
@@ -376,32 +338,26 @@ def on_trial_result(self, trial: Trial, result: dict) -> str:
         self._check_result(result)
         trial_id = str(trial.trial_id)
         trial_decision = SchedulerDecision.CONTINUE
-        if len(result) == 0:
```
This cannot happen. It is some old code, when the scheduler was used with another backend (AutoGluon), where this could happen.

jgolebiowski replied:
Ok makes sense, thanks for clearing it up.

mseeger commented on syne_tune/optimizer/schedulers/searchers/bore/bore.py:
```diff
@@ -37,19 +36,18 @@ def __init__(
         self,
         config_space: dict,
         metric: str,
-        points_to_evaluate=None,
-        mode: str = "max",
-        gamma: float = 0.25,
-        calibrate: bool = False,
-        classifier: str = "xgboost",
-        acq_optimizer: str = "rs",
-        feval_acq: int = 500,
-        random_prob: float = 0.0,
-        init_random: int = 6,
-        classifier_kwargs: dict = None,
```
That is best practice for methods of base classes. We can state the default values in the docstring.

Also, if the assignments of default values happen just at the start of a method, a user looking at the code will have no more trouble seeing them.

jgolebiowski replied:
I see, as long as the default values are in the docstring it makes sense to me.

mseeger replied:
OK, will do that in second PR, where docstrings are updated.

mseeger commented on syne_tune/optimizer/schedulers/searchers/bore/multi_fidelity_bore.py:
```diff
@@ -10,72 +10,63 @@
-        feval_acq: int = 500,
-        random_prob: float = 0.0,
-        init_random: int = 6,
-        classifier_kwargs: dict = None,
+        points_to_evaluate: Optional[List[dict]] = None,
+        mode: Optional[str] = None,
+        gamma: Optional[float] = None,
+        calibrate: Optional[bool] = None,
+        classifier: Optional[str] = None,
+        acq_optimizer: Optional[str] = None,
+        feval_acq: Optional[int] = None,
+        random_prob: Optional[float] = None,
+        init_random: Optional[int] = None,
+        classifier_kwargs: Optional[dict] = None,
         resource_attr: str = "epoch",
-        **kwargs
```
That is the whole point in the subclass. If something depends to the base class, you should *have* to look there.
Copy and paste is not the solution.

I am all for mentioning default values in the docstring. This is apparently done in Google style, but not in reST we are using. I am not an expert on this.

It is even worse if we end up relying on copy and paste, and then suddenly somebody changes a def value in a derived class. Is this a copy mistake, or is it intended?

jgolebiowski replied:
I see, as long as the default values are in the docstring it makes sense to me.

mseeger commented:
OK, I think I have addressed concerns. About default values: We can include them in docstrings, but I am against copy
and paste them to subclasses. It is best practice to use `= None` and deal with them in the code of the base class.

jgolebiowski commented:
Good changes - I agree with the current implementation if we add defaults into docstrings for classes that are inherited from.

jgolebiowski reviewed (approved):
Docsting changes will come in the 2nd PR as agreed offline.
